doc/gnucash-cli.1.in
.\" Written by Robert Merkel (rgmerk@mira.net)
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH GNUCASH 1 "@GNC_VCS_REV_DATE@" "Version @GNC_VCS_REV@" "GnuCash Manual Pages"
.SH NAME
gnucash-cli \- command line tool to go with GnuCash
.SH SYNOPSIS
.B gnucash-cli [options] [
.I accounts.gnucash
.B ]
.SH DESCRIPTION
.B gnucash-cli
enables certain intraction with a gnucash datafile directly from
the command line.

It has two modes:
.B quotes
mode and
.B report
mode.

.SH Quotes Mode (activated with --quotes <cmd>)
This mode has options to get online quotes for foreign currencies and stocks
in the given data file. It supports the following command:
.IP info
Shows Finance::Quote version and exposed quote sources.
.IP get
Retrieves online quotes. Note this requires Finance::Quote to be properly
configured for your system.

The
.B get
command takes the following options:
.IP --namespace=REGEXP
Regular expression determining which namespace commodities will be retrieved.

.IP dump
Fetches current quotes for specified currencies or stocks from a specified
namespace and print the results to the console. This must be followed with
a source and one or more symbols, unless the source is "currency" in which
case it must be followed with two or more symbols, the first of which is
the currency in which exchange rates for the rest will be quoted.

The
.B dump
command takes the option
.IP -v
.IP --verbose
Lists all of the parameters Finance::Quote returns for the symbol instead of
only the ones that Gnucash requires.

.SH Report Mode (activated with --report <cmd>)
This mode has options to work with reports in the given data file.
It supports the following command:
.IP run
Runs a report on the given data file.

The
.B run
command takes the following options:
.IP --name=REPORT_NAME
Name of the report to run
.IP --export-type=TYPE
Specify export type
.SH General Options
.IP --version
Show
.B GnuCash
version
.IP --help
Show this help message
.IP --debug
Enable debugging mode: increasing logging to provide deep detail.
.IP --extra
Enable extra/development/debugging features.
.IP --log
Log level overrides, of the form "log.ger.path={debug,info,warn,crit,error}"
This option can be specified multiple times.
.IP --logto
File to log into; defaults to "$TMPDIR/gnucash.trace"; use "stderr" or "stdout"
to write to the terminal.
.SH FILES
.I ~/.gnucash/config.auto
.RS
Automatically generated per-user configuration file.  DO NOT EDIT
MANUALLY!
.SH ENVIRONMENT
.IP GNC_DEBUG
Enable debugging output.  This allows you to turn on the debugging
earlier in the startup process than you can with
.B --debug.
.IP GUILE_LOAD_PATH
An override for the
.B GnuCash
scheme load path. You shouldn't need to alter this.  For more
information see the README file.
.IP GNC_MODULE_PATH
An override for the
.B GnuCash
scheme load path for
.B Gnucash
modules. You shouldn't need to alter this.  For more
information see the README file.
.SH BUGS
.B Please report any bugs using the bug reporting form on the
.B GnuCash
web site.

.SH "SEE ALSO"
.B GnuCash
has extensive on-line help available from the "help" menu.

Additional information and the latest version is available
at the web site:
.B https://www.gnucash.org

.SH AUTHOR
Robert Merkel <rgmerk@mira.net> wrote this manpage.  

.B GnuCash
was written by Robin Clark, Linas Vepstas, and a cast of thousands.


NEWS
Version history:
------- -------
5.0  - 26 March 2023

Between 4.904 and 5.0, the following bugfixes were accomplished:
    Bug 798792 - "Delete Split" missing from context menu
    Bug 798793 - Cursor jumps to the end of the field.

The following fixes and improvements were not associated with bug reports:
    Fix crash in gnc_owner_get_owner
    [gnc-account-sel] fix entered_text leak
    Fix missing statusbar tooltips for popup menus
    Popup menu missing from sx template transactions window

New and Updated Translations: Chinese (Simplified), Croatian, Hungarian,
    Japanese, Portuguese, Portuguese (Brazil), Spanish, Swedish

4.14 - 26 March 2023

Between 4.13 and 4.14, the following bugfixes were accomplished:

    Bug 797477 - Manual foreign transaction from APAR to another doesn't
        trigger price input
    Bug 797725 - Untranslatable string "For Period Covering ~a to ~a"
    Bug 797903 - Transaction window: The 'blue line' is always (erroneously)
        topmost when t. sorting order is inverted.
    Bug 798570 - Totals for the income, expenses and remaining to budget
        incorrect for one specific date.
    Bug 798625 - "Last up through report date" changed in 4.12
                 Don't adjust report dates with time64CanonicalDayTime. They're
                 set to 23:59:59 local which is always >= neutral time at which
                 prices and transactions are timestamped. Local noon is before
                 neutral time for all time zones east of UTC+2.
    Bug 798679 - Unicode normalization should be used for comparison but not
        stored.
                 Change to NFC normalization for all comparisons because the
                 Unicode meaning of compatible might collide with the user's
                 intent.
    Bug 798702 - Crash in gnc_plugin_page_focus_idle_destroy() closing a report
        before it completes.
    Bug 798705 - ] New: UI string mismatch: OK vs. Next
    Bug 798712 - Regional Decimal Point not recognized in budgets
    Bug 798717 - Reports > Business > Fancy Invoice duplicates company details
    Bug 798732 - : Type of balance carry forward accounts for german skr49
        account template fixed
    Bug 798734 - Aging Reports don't handle mixed currency payments and
        invoices without Trading Accounts
    Bug 798737 - Minor grammatical error - 'for' missing in 'You will be asked
        a conversion rate for each.'
    Bug 798740 - Build fails with gcc 13
    Bug 798747 - Crash in Investment Portfolio report
    Bug 798748 - Transaction Notes field's value does not appear in reverse
        transaction.
    Bug 798754 - Build fails with gcc 13 and glib > 2.76
    Bug 798759 - Register widths for 'with sub-account' not saved
    Bug 798765 - Increase the limit for Maximum Slices in the Expense Piechart

The following fixes and improvements were not associated with bug reports:

    Budget totals for one period would be wrong when increasing the number of
        periods.
    Allow C decls after statements, legal since C99.
    [kvp] Use static strings and boxed in gvalue_from_kvp_value. Saves
        allocating and copying complex values, avoiding potential memory leaks.
    L10N:DE:Updating accounts from 0001 to 8000 DATEV-SKR04 2022-01-01.
    Payment dialog - always list all possible post accounts. Before only the
        post account in the currency of the selected vendor or customer was
        available.
    Revisit invoice payment in multi-currency context:
        * Show proper amount in dialog when applying or editing an existing
          transaction as payment
          user chose a new transfer account and the old currency is neither the
          new transfer account's currency nor the post account's currency)
    Edit/assign payment - be more prudent not to lose data
    We can't handle the case where a transaction has splits in more than one
        APAR account. Instead of only warning that some of them will be lost
        refuse to continue and leave it to the user to fix the transaction
        first.
    Multi-currency - show split amount rather than transaction value
    Activate feature GNC_FEATURE_EQUITY_TYPE_OPENING_BALANCE
    [balsheet-pnl] move out of experimental into regular menu. After 4 years it
        can finally graduate.
    [dialog-sx-editor]Allow a daily SX to show all recurrences in the calendar
        instead of only 6 months.
    Update Form/Schedule line references for 2022 for the US Income Tax Report

New and Updated Translations: Chinese (Simplified), Croatian, Czech,
    English (Australia), English (New Zealand), English (United Kingdom),
    French, Hebrew, Hungarian, Italian, Japanese, Macedonian, Polish,
    Portuguese, Portuguese (Brazil), Russian, Spanish, Swedish, Ukrainian

4.904 - 19 March 2023
Between 4.903 and 4.904, the following bugfixes were accomplished:
The following fix will also appear in GnuCash 4.14:

    Bug 798570 - Budget totals for income, expenses and remaining to budget
                 incorrect for one specific date.

The following fixes are in 4.904 only and will not appear in GnuCash 4.14:

    Bug 798558 - Irregular line spacing in Account Tree View
    Bug 798773 - Help file launches in browser when using command-H to hide
                 GnuCash
                 This is a MacOS-only problem. Note that for the fix to have
                 effect you must remove the file accelerator-map from
                 GNC_CONFIG_HOME when GnuCash isn't running.
    Bug 798778 - GnuCash quits abruptly when attempting to edit options for
                 certain reports.
    Bug 798781 - Printing multiple checks prints duplicate checks
    Bug 798787 - gnc-register-owner-option currently broken

The following fixes and improvements were not associated with bug reports:

    Fix quotes codecvt to wchar_t error on Windows.
    Budget totals for one period would be wrong when increasing the number of
        periods.

gnucash/gnucash-cli.cpp
/*
 * gnucash-cli.cpp -- The command line entry point for GnuCash
 *
 * Copyright (C) 2020 Geert Janssens <geert@kobaltwit.be>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, contact:
 *
 * Free Software Foundation           Voice:  +1-617-542-5942
 * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652
 * Boston, MA  02110-1301,  USA       gnu@gnu.org
 */
#include <config.h>

#ifdef __MINGW32__
#include <Windows.h>
#include <fcntl.h>
#endif

#include "gnucash-commands.hpp"
#include "gnucash-core-app.hpp"

#include <glib/gi18n.h>
#include <gnc-engine.h>
#include <gnc-prefs.h>

#include <boost/locale.hpp>
#include <boost/optional.hpp>
#ifdef __MINGW32__
#include <boost/nowide/args.hpp>
#endif
#include <iostream>
#include <gnc-quotes.hpp>

namespace bl = boost::locale;

/* This static indicates the debugging module that this .o belongs to.  */
[[maybe_unused]] static QofLogModule log_module = GNC_MOD_GUI;

namespace Gnucash {

    class GnucashCli : public CoreApp
    {
    public:
        GnucashCli (const char* app_name);
        void parse_command_line (int argc, char **argv);
        int start (int argc, char **argv);
    private:
        void configure_program_options (void);

        std::vector<std::string> m_quotes_cmd;
        boost::optional <std::string> m_namespace;
        bool m_verbose = false;

        boost::optional <std::string> m_report_cmd;
        boost::optional <std::string> m_report_name;
        boost::optional <std::string> m_export_type;
        boost::optional <std::string> m_output_file;
    };

}

Gnucash::GnucashCli::GnucashCli (const char *app_name) : Gnucash::CoreApp (app_name)
{
    configure_program_options();
}

void
Gnucash::GnucashCli::parse_command_line (int argc, char **argv)
{
    Gnucash::CoreApp::parse_command_line (argc, argv);

    if (!m_log_to_filename || m_log_to_filename->empty())
        m_log_to_filename = "stderr";

    if (m_namespace)
        gnc_prefs_set_namespace_regexp (m_namespace->c_str());
}

// Define command line options specific to gnucash-cli.
void
Gnucash::GnucashCli::configure_program_options (void)
{
    bpo::options_description quotes_options(_("Price Quotes Retrieval Options"));
    quotes_options.add_options()
        ("quotes,Q", bpo::value<std::vector<std::string>> (&m_quotes_cmd)->multitoken(),
     _("Execute price quote related commands. The following commands are supported.\n\n"
       "  info: \tShow Finance::Quote version and exposed quote sources.\n"
       "   get: \tFetch current quotes for all foreign currencies and stocks in the given GnuCash datafile.\n"
       "  dump: \tFetch current quotes for specified currencies or stocks from a specified namespace and print the results to the console.\n"
       "        \tThis must be followed with a source and one or more symbols, unless the source is \"currency\" in which case it must be followed with two or more symbols, the first of which is the currency in which exchange rates for the rest will be quoted.\n"))
    ("namespace", bpo::value (&m_namespace),
     _("Regular expression determining which namespace commodities will be retrieved for when using the get command"))
     ("verbose,V", bpo::bool_switch (&m_verbose),
      _("When using the dump command list all of the parameters Finance::Quote returns for the symbol instead of the ones that Gnucash requires."));

    m_opt_desc_display->add (quotes_options);
    m_opt_desc_all.add (quotes_options);

    bpo::options_description report_options(_("Report Generation Options"));
    report_options.add_options()
    ("report,R", bpo::value (&m_report_cmd),
     _("Execute report related commands. The following commands are supported.\n\n"
     "  list: \tLists available reports.\n"
     "  show: \tDescribe the options modified in the named report. A datafile \
may be specified to describe some saved options.\n"
     "  run: \tRun the named report in the given GnuCash datafile.\n"))
    ("name", bpo::value (&m_report_name),
     _("Name of the report to run\n"))
    ("export-type", bpo::value (&m_export_type),
     _("Specify export type\n"))
    ("output-file", bpo::value (&m_output_file),
     _("Output file for report\n"));
    m_opt_desc_display->add (report_options);
    m_opt_desc_all.add (report_options);

}

int
Gnucash::GnucashCli::start ([[maybe_unused]] int argc, [[maybe_unused]] char **argv)
{
    Gnucash::CoreApp::start();

    if (!m_quotes_cmd.empty())
    {
        if (m_quotes_cmd.front() == "info")
        {
            return Gnucash::check_finance_quote ();
        }
        else if (m_quotes_cmd.front() == "get")
        {
            if (!m_file_to_load && m_quotes_cmd.size() == 2)
                m_file_to_load = m_quotes_cmd[1];

            if (!m_file_to_load || m_file_to_load->empty())
            {
                std::cerr << bl::translate("Missing data file parameter") << "\n\n"
                << *m_opt_desc_display.get() << std::endl;
                return 1;
            }
            else
                return Gnucash::add_quotes (m_file_to_load);
        }
        else if (m_quotes_cmd.front() == "dump")
        {
            if (m_quotes_cmd.size() < 3 ||
                (m_quotes_cmd[1] == "currency" &&
                 m_quotes_cmd.size() < 4))
            {
                std::cerr << bl::translate("Not enough information for quotes dump") << std::endl;
                return 1;
            }
            auto source = m_quotes_cmd[1];
            m_quotes_cmd.erase(m_quotes_cmd.begin(), m_quotes_cmd.begin() + 2);
            return Gnucash::report_quotes(source.c_str(), m_quotes_cmd,
                                          m_verbose);
        }
        else
        {
            std::cerr << bl::format (bl::translate("Unknown quotes command '{1}'")) % m_quotes_cmd.front() << "\n\n"
                      << *m_opt_desc_display.get() << std::endl;
            return 1;
        }
    }

    if (m_report_cmd)
    {
        if (*m_report_cmd == "run")
        {
            if (!m_file_to_load || m_file_to_load->empty())
            {
                std::cerr << _("Missing data file parameter") << "\n\n"
                          << *m_opt_desc_display.get() << std::endl;
                return 1;
            }
            else
                return Gnucash::run_report(m_file_to_load, m_report_name,
                                           m_export_type, m_output_file);
        }

        // The command "list" does *not* test&pass the m_file_to_load
        // argument because the reports are global rather than
        // per-file objects. In the future, saved reports may be saved
        // into the datafile, therefore one will need to be specified
        // for loading.
        else if (*m_report_cmd == "list")
                return Gnucash::report_list ();

        // The command "show" does test&pass the m_file_to_load
        // argument, and will attempt to load datafile prior to
        // describing report. If loading fails, it will continue
        // showing report options.
        else if (*m_report_cmd == "show")
            if (!m_report_name || m_report_name->empty())
            {
                std::cerr << _("Missing --name parameter") << "\n\n"
                          << *m_opt_desc_display.get() << std::endl;
                return 1;
            }
            else
                return Gnucash::report_show (m_file_to_load, m_report_name);
        else
        {
            std::cerr << bl::format (std::string{_("Unknown report command '{1}'")}) % *m_report_cmd << "\n\n"
                      << *m_opt_desc_display.get();
            return 1;
        }
    }

    std::cerr << _("Missing command or option") << "\n\n"
              << *m_opt_desc_display.get() << std::endl;

    return 1;
}

int
main(int argc, char **argv)
{
    const char *app_name = PROJECT_NAME "-cli";
    Gnucash::GnucashCli application (app_name);
#ifdef __MINGW32__
    boost::nowide::args a(argc, argv); // Fix arguments - make them UTF-8
#endif
    application.parse_command_line (argc, argv);
    application.start (argc, argv);

    return 0;
}
